### PR TITLE
NAS-3681: PDM management including scan model/schemata

### PR DESCRIPTION
**Pull request by jaceksan**

### CODE REVIEW

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

File: gooddata-sdk/gooddata_sdk/catalog/data_source/service.py [hunk 1]
```diff
@@ -148,3 +155,58 @@ def load_and_put_declarative_data_sources(
...
+        self._layout_api.set_pdm_layout(data_source_id, declarative_pdm)
+
+    def scan_data_source(
+        self, data_source_id: str, scan_request: CatalogScanModelRequest = CatalogScanModelRequest()
+    ) -> CatalogDeclarativeTables:
+        # TODO - how to process warnings? Return it to developers, so they can process it in their way?
+        scan_response = self._scan_api.scan_pdm(data_source_id, scan_request.to_api())
+        warnings = scan_response.get("warnings")
+        if warnings:
+            print("Scan produced the following warnings: ")
```
Comment (pcerny) [inline gooddata-sdk/gooddata_sdk/catalog/data_source/service.py:173]: Printing warnings from library is not a good idea:
- it is not possible to block writing messages to output
- user cannot process warning automatically

Lets return raw warnings together with CatalogDeclarativeTables for now. We can provide some method to print data, but it must be called explicitly.

File: gooddata-sdk/gooddata_sdk/catalog/data_source/service.py [hunk 2]
```diff
@@ -148,3 +155,58 @@ def load_and_put_declarative_data_sources(
...
+        self, data_source_id: str, scan_request: CatalogScanModelRequest = CatalogScanModelRequest()
+    ) -> None:
+        self.put_declarative_pdm(data_source_id, self.scan_data_source(data_source_id, scan_request))
+
+    def data_source_folder(self, data_source_id: str, layout_root_path: Path) -> Path:
+        layout_organization_folder = self.layout_organization_folder(layout_root_path)
+        data_sources_folder = CatalogDeclarativeDataSources.data_sources_folder(layout_organization_folder)
+        return CatalogDeclarativeDataSource.data_source_folder(data_sources_folder, data_source_id)
+
+    def store_declarative_pdm(self, data_source_id: str, layout_root_path: Path = Path(os.path.curdir)) -> None:
```
Comment (pcerny) [inline gooddata-sdk/gooddata_sdk/catalog/data_source/service.py:197]: Replace `Path(os.path.curdir)` with `Path.cwd()`

File: gooddata-sdk/gooddata_sdk/catalog/data_source/service.py [hunk 3]
```diff
@@ -148,3 +155,58 @@ def load_and_put_declarative_data_sources(
...
+        data_source_folder = self.data_source_folder(data_source_id, layout_root_path)
+        self.get_declarative_pdm(data_source_id).store_to_disk(data_source_folder)
+
+    def load_declarative_pdm(
+        self, data_source_id: str, layout_root_path: Path = Path(os.path.curdir)
```
Comment (pcerny) [inline gooddata-sdk/gooddata_sdk/catalog/data_source/service.py:202]: Replace `Path(os.path.curdir)` with `Path.cwd()`

File: gooddata-sdk/gooddata_sdk/catalog/data_source/service.py [hunk 4]
```diff
@@ -148,3 +155,58 @@ def load_and_put_declarative_data_sources(
...
+    ) -> CatalogDeclarativeTables:
+        data_source_folder = self.data_source_folder(data_source_id, layout_root_path)
+        return CatalogDeclarativeTables.from_dict(CatalogDeclarativeTables.load_from_disk(data_source_folder))
+
+    def load_and_put_declarative_pdm(self, data_source_id: str, layout_root_path: Path = Path(os.path.curdir)) -> None:
```
Comment (pcerny) [inline gooddata-sdk/gooddata_sdk/catalog/data_source/service.py:207]: Replace `Path(os.path.curdir)` with `Path.cwd()`

File: gooddata-sdk/tests/catalog/test_catalog_data_source.py [hunk 5]
```diff
@@ -415,3 +416,93 @@ def test_declarative_data_sources(test_config):
...
+
+    pdm = sdk.catalog_data_source.scan_data_source(data_source_id)
+    assert len(pdm.tables) == 5
+
+    scan_request = CatalogScanModelRequest(scan_tables=False, scan_views=True)
+    pdm = sdk.catalog_data_source.scan_data_source(data_source_id, scan_request)
+    assert len(pdm.tables) == 0
+
+    try:
+        CatalogScanModelRequest(scan_tables=False, scan_views=False)
```
Comment (pcerny) [inline gooddata-sdk/tests/catalog/test_catalog_data_source.py:457]: Use
```
with pytest.raises(ValueError)
    CatalogScanModelRequest(scan_tables=False, scan_views=False)
```

**Review by pcerny — CHANGES_REQUESTED** [review]

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

File: gooddata-sdk/gooddata_sdk/catalog/data_source/service.py (hunk 1 above)
Comment (jaceksan) [reply gooddata-sdk/gooddata_sdk/catalog/data_source/service.py:169]: warnings printed only on demand.
ScanResult is introduced as a standard Catalog class containing both PDM and warnings.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

**Review by pcerny — APPROVED** [review]

**Comment by yenkins** [comment]
### Sonar scan result


More detail, see in https://sonarqube-gate.intgdc.com/dashboard?id=gooddata-python-sdk-gate-PR94

To scan for vulnerabilities in dependencies and run unit tests (to get coverage report in sonar) please comment your PR with 'extended check sonar'.

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-python-sdk-sonarqube-scan-zuul-gate/86/ : SUCCESS in 54s
- https://checklist.intgdc.com/job/compliance/job/gooddata-python-sdk-fossa-scanning-zuul-el7/86/ : SUCCESS in 2m 22s (non-voting)
- https://checklist.intgdc.com/job/compliance/job/gooddata-python-sdk-merge-compliance-check/86/ : SUCCESS in 19s

**Pull request merged**